### DISCUSSION TITLE
Linearity of the Austral.Memory Module

### DISCUSSION THREAD

**Discussion by bwburnsides** [austral/austral#606] [General]
I've recently discovered Austral and it's been a great resource for learning about linear types, which is a capability I think I want bake into a language I'm designing; alongside second class references, they seem like a solution to memory safety that doesn't "break the bank" so to speak.

However something surprising to me is that the `Pointer` type in the builtin memory module is a `Free` type. This seems a bit counterintuitive to me, especially since the write-up on memory safety presents linear pointers as a use-case for linear types.

What's the rationale here? Is it because pointers can be used in Austral to point to non-dynamic memory? (stack, static)

My thoughts are that assignment would be an illegal operation on pointers such that once allocated (via a library call), the pointer can't be used to point to a different element - such as one on a stack. In that sense pointers would be used specifically for dynamically allocated resources. (Come to think of it, if pointer was linear, wouldn't that naturally prevent assignment?)

I'm just curious to learn more about this decision in hopes that it'll fill in some of my linear type knowledge gaps.

**Comment by eudoxia0** [comment]
The rationale is that linear APIs can be built (and, in most cases, necessarily have to be built) on an unsafe foundation.

e.g., a linear `File` handle type internally contains a C-style `FILE*` unsafe file pointer. Any linear API that wraps a foreign library has to contain unsafe FFI internals. A linear API to e.g. SQLite has to contain unsafe SQLite library types. And so on.

Memory is just another case: the linear APIs in the standard library (the `Box` type, but also collections like `Buffer`) provide safe, linear interfaces on top of unsafe memory management.
